### ISSUE TITLE
Admin shows thousands of cart items for customers

### ISSUE THREAD

**Issue by xtremevision** [magento/magento2#27564]
### Preconditions (*)
1. Centos OS 7 x64
2. Magento 2.3.3
3. PHP 7.2

### Steps to reproduce (*)
1. Create lots of carts / quotes, both as logged in user and guest
2. Keep one customer without any quotes
3. Goto backend, edit customer, click on Shopping Cart tab
4. Grid shows thousand of items

### Expected result (*)
1. Shopping cart tab should show customer's cart items if available, or none if no carts available for customer

### Actual result (*)
I am experiencing this issue in the backend admin, in the customer record, Shopping Cart tab. It looks like the getQuote() function returns all cart items (without a parent id), if no active quote is found for customer id [X].

See screenshot:

![2020-04-02 21 40 21](https://user-images.githubusercontent.com/6533394/78367871-c0496f80-7590-11ea-9e03-3499be0a4dbd.png)

This getQuote() function is at fault from what I can tell.

![2020-04-02 21 42 05](https://user-images.githubusercontent.com/6533394/78367877-c3446000-7590-11ea-838f-d5a92b8a48ca.png)

The quoteRepository->getForCustomer() function throws the exception as shown here:

![2020-04-02 21 45 26](https://user-images.githubusercontent.com/6533394/78367885-c5a6ba00-7590-11ea-9576-0d0d3d5db508.png)

The trouble is if no quote is found, then _prepareCollection() will return all quote_items regardless of quote id, regardless of customer_id, as shown by this query:

SELECT main_table.* FROM quote_item AS main_table WHERE (parent_item_id IS NULL)

I fixed in by adding this check:

```
protected function _prepareCollection()
{
$quote = $this->getQuote();

    if ($quote) {
        $collection = $quote->getItemsCollection(false);
    } else {
        $collection = $this->_dataCollectionFactory->create();
    }

    $collection->addFieldToFilter('parent_item_id', ['null' => true]);

    /**** Added by me as temporary fix ***/
    if(is_null($quote->getId()))
        $collection->addFieldToFilter('quote_id', -1);
    /**** Added by me as temporary fix ***/

    $this->setCollection($collection);

    return parent::_prepareCollection();
}

```
It probably isn't the best or most elegant fix, but it works, or so it seems.
Does anyone have a better idea?

Thanks,
Michael Mussulis

**Comment by krishprakash** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @krishprakash. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @krishprakash, here is your [Magento instance](https://i-27564-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-27564-2-4-develop.instances.magento-community.engineering/admin_cbb5](https://i-27564-2-4-develop.instances.magento-community.engineering/admin_cbb5)
Login: `043c35ee` Password: `f3e2f039d232`
Instance will be terminated in up to 3 hours.

**Comment by krishprakash** [comment]
Hi @xtremevision. Thank you for your report.
Not able to Reproduce the issue on latest mangeto 2.3.4 instance 
Kindly provide more information 
Also please confirm the issue on latest 2.3.4 instance
Thanks !!

**Comment by xtremevision** [comment]
Hi @krishprakash,

Did you create lots of quotes? What does your quote table look like?

Thanks,
Michael.

**Comment by krishprakash** [comment]
Hi @xtremevision. Yes created many quotes for login and guest user as well
We can not see quote table for provided instance.
So if possible please share the video with steps you mentioned so will review in local instance
Thank You !!

**Comment by xtremevision** [comment]
Hi @krishprakash ,

I've got it figured out. The quote table was missing records for customers. This was causing the bad view in the admin, shopping cart tab.

I used a module called Login as Customer, to login as the customer, and this resulted in an empty quote record being created for that customer. Checking in the backend in the Shopping Cart tab now does not show any cart items for the customer.

It seems to me that Magento's logic is that each customer MUST have at least 1 empty cart in order for the Shopping Cart grid to function properly, and to avoid the No such entity with customerId = XXX exception.

**Comment by krishprakash** [comment]
@xtremevision. Then this is the compatibility issue with your custom extension not the default magento issue

**Comment by krishprakash** [comment]
As this compatibility issue with your custom extension so closing this
Thanks !!

**Comment by xtremevision** [comment]
No, I don't agree. This is has nothing to do with a my custom extension. Why did you close this? The module I used was just to prove the fix. Please reopen.

**Comment by codercv** [comment]
I have same issue on 2.3.3-p1 migrated from 1.9.3.1. Thank you @xtremevision for you quick fix.

**Comment by JKetelaar** [comment]
Same issue here

**Comment by JKetelaar** [comment]
@xtremevision could you provide some more information about your debug and what seems to be the issue with "empty carts"?

**Comment by JKetelaar** [comment]
Are you talking about the result of:
```
SELECT customer_entity.entity_id FROM customer_entity LEFT JOIN quote ON quote.customer_id = customer_entity.entity_id WHERE quote.entity_id IS NULL;
```

**Comment by nuzil** [comment]
An confirm, issue is there on 2.3.5-p1

**Comment by portchris** [comment]
I had to tweak @xtremevision's manual fix a tad to for my particular version (2.3.3). 

di.xml
```
    <!-- 
    /**
      * Core bug fix 
      * @see    https://github.com/magento/magento2/issues/27564 
      */
    -->
    <preference for="Magento\Customer\Block\Adminhtml\Edit\Tab\Cart" type="Convert\CustomerContent\Block\Adminhtml\Edit\Tab\Cart" />

```

app/code/MyVendor/MyModule/Block/Adminhtml/Edit/Tab/Cart.php
```
/**
 * @author  YOUR NAME
 * @since   2020-08-13 (Magento 2.3.3) <1.0.0>
 * @see     https://github.com/magento/magento2/issues/27564 
 */

namespace MyVendor\MyModule\Block\Adminhtml\Edit\Tab;

use Magento\Customer\Block\Adminhtml\Edit\Tab\Cart as Original;

class Cart extends Original
{
    protected function _prepareCollection()
    {
        parent::_prepareCollection();

        $quote = $this->getQuote();
        if ($quote) {
            $collection = $quote->getItemsCollection(false);
        } else {
            $collection = $this->_dataCollectionFactory->create();
        }

        $collection->addFieldToFilter('parent_item_id', ['null' => true]);

        /** BEGIN: fix */
        if (is_null($quote->getId()))
            $collection->addFieldToFilter('quote_id', -1);
        /** END: fix */

        $this->setCollection($collection);

        return $this;
    }
}
```

**Comment by WinstonN** [comment]
We see this as well, Magento 2.3.5-p2

**Comment by Rich07082** [comment]
same here, same issue in 2.3.5-p2

**Comment by Rich07082** [comment]
Just confirmed what xtremevision said, if customer doesn't have a valid open cart in the quote table it will list everything that is open.  I check two new account opened one did have an open cart in the quote table, worked as it should.  No items listed.  Looked at another customer, not open quotes and listed everything.
Rich

**Comment by JeroenBoersma** [comment]
FYI: This is fixed in Magento 2.4 and up...

Magento backend will alsways return a quote(even a empty one)
Using this items collection will not apply the quoteId as a filter when searching/loading cart items resulting in loading all cart items for any customer.

In 2.4 the `->getId()` is applied again and the a empty collection is returned, which is indeed the correct behavior.

**Comment by kzobov** [comment]
experiencing this in 2.3.5-p2

**Comment by alinolandry** [comment]
I have a same issue in 2.3.5-p2

**Comment by Vishrootways** [comment]
I have the same issue in Magento commerce cloud 2.3.4, and this has been solved in Magento 2.4.3.

But I am having one more issue: sometimes front-end shopping cart items get duplicated, too.
E.g., customers added 10 items to the cart, and when proceeding with the order, there will be 20 items in the order. 
Does anyone have this issue?

**Comment by sukeshini-kot** [comment]
In Magento 2.4 it has been fixed.
FILE : vendor/magento/module-customer/Block/Adminhtml/Edit/Tab/Cart.php

`/**
     * Prepare collection
     *
     * @return $this
     */
    protected function _prepareCollection()
    {
        $quote = $this->getQuote();

        if ($quote && $quote->getId()) {
            $collection = $quote->getItemsCollection(false);
            $collection->addFieldToFilter('parent_item_id', ['null' => true]);
        } else {
            $collection = $this->_dataCollectionFactory->create();
        }

        $this->setCollection($collection);

        return parent::_prepareCollection();
    }`
